src/postprocessing.py

```diff
@@ -1,386 +1,38 @@
-    mask_features = {'iou': iou, 'threshold': threshold, 'area': area, 'mean_prob': mean_prob,
-                     'max_prob': max_prob, 'bbox_ar': bbox_aspect_ratio,
-                     'bbox_area': bbox_area, 'bbox_fill': bbox_fill, 'min_dist_to_border': min_dist_to_border,
-                     'max_dist_to_border': max_dist_to_border, 'contour_length': contour_length}
-    return mask_features
-
-
-def get_iou_matrix(labels, annotations):
-    mask_anns = []
-    if annotations is None or annotations == []:
-        return None
-    else:
-        for annotation in annotations:
-            if not isinstance(annotation['segmentation'], dict):
-                annotation['segmentation'] = \
-                    cocomask.frPyObjects(annotation['segmentation'], labels.shape[0], labels.shape[1])[0]
-        annotations = [annotation['segmentation'] for annotation in annotations]
-        for label_nr in range(1, labels.max() + 1):
-            mask = labels == label_nr
-            mask_ann = rle_from_binary(mask.astype('uint8'))
-            mask_anns.append(mask_ann)
-        iou_matrix = cocomask.iou(mask_anns, annotations, [0, ] * len(annotations))
-        return iou_matrix
-
-
-def get_iou(iou_matrix, label_nr):
-    if iou_matrix is not None:
-        return iou_matrix[label_nr - 1].max()
-    else:
-        return None
-
-
-def get_thresholds():
-    thresholds = []
-    for n_thresholds in CATEGORY_LAYERS:
-        threshold_step = 1. / (n_thresholds + 1)
-        category_thresholds = np.arange(threshold_step, 1, threshold_step)
-        thresholds.extend(category_thresholds)
-    return thresholds
-
-
-def get_bbox(mask):
-    '''taken from https://stackoverflow.com/questions/31400769/bounding-box-of-numpy-array and
-    modified to prevent bbox of zero area'''
-    rows = np.any(mask, axis=1)
-    cols = np.any(mask, axis=0)
-    rmin, rmax = np.where(rows)[0][[0, -1]]
-    cmin, cmax = np.where(cols)[0][[0, -1]]
-    return rmin, rmax + 1, cmin, cmax + 1
-
-
-def get_min_max_distance_to_border(bbox, im_size):
-    min_distance = min(bbox[0], im_size[0] - bbox[1], bbox[2], im_size[1] - bbox[3])
-    max_distance = max(bbox[0], im_size[0] - bbox[1], bbox[2], im_size[1] - bbox[3])
-    return min_distance, max_distance
 
 
-def get_contour(mask):
-    mask_contour = np.zeros_like(mask).astype(np.uint8)
-    _, contours, hierarchy = cv2.findContours(mask.astype(np.uint8), cv2.RETR_TREE, cv2.CHAIN_APPROX_NONE)
-    cv2.drawContours(mask_contour, contours, -1, (255, 255, 255), 1)
-    return mask_contour
+class SubmissionProducer(BaseTransformer):
 
+    def transform(self, image_ids, results, image_size, decoder_dict):
+        self.decoder_dict = decoder_dict
+        prediction_strings = []
+        for bboxes, labels in results:
+            prediction_strings.append(self.get_prediction_string(bboxes, labels, image_size))
+        submission = pd.DataFrame({'ImageId': image_ids, 'PredictionString': prediction_strings})
```

Comment:
@apyskir I would expect to see '_'.join(prediction_strings) at some point because it seems to be passing just a list here. Am I right ?

Reply:
@jakubczakon No, you're not right. `prediction_string` is a list of strings, each string for a separate image. You don't want to join them. Each of them should be in a separate line in output file.